### PR TITLE
[Merged by Bors] - feat(Algebra/Order/Ring/WithTop): add generalised and specialised power of top 

### PR DESCRIPTION
**Pull request by pitmonticone**
- [x] Add generalised version of `top_pow` in `Algebra/Order/Ring/WithTop`
- [x] Add specialised version of `top_pow` in `Data/ENat/Basic.lean`
- [x] Add specialised version of `top_pow` in `Data/ENNReal/Operations.lean` 

Co-authored by @D-Thomine.

### CODE REVIEW

File: Mathlib/Data/ENat/Basic.lean [hunk 1]
```diff
@@ -89,6 +89,16 @@ theorem coe_sub (m n : ℕ) : ↑(m - n) = (m - n : ℕ∞) :=
...
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℕ∞)^n = ⊤ := by
+  apply @Nat.le_induction 1 (fun m : ℕ ↦ fun _ : 1 ≤ m ↦ (⊤ : ℕ∞) ^ m = ⊤) (pow_one ⊤)
+  · intro m _ h
+    calc
+      (⊤ : ℕ∞)^(m + 1) = ⊤^m * ⊤^1 := by rw [pow_add ⊤ m 1]
+                     _ = ⊤ * ⊤^1   := by rw [h]
+                     _ = ⊤ * ⊤     := by rw [pow_one ⊤]
+                     _ = ⊤         := WithTop.top_mul_top
+  · exact n_pos
```
Comment (sgouezel) [inline Mathlib/Data/ENat/Basic.lean:100]: When the steps of the computation are trivial, it's often better to just use `simp`, like
```suggestion
  apply Nat.le_induction (P := fun m : ℕ ↦ fun _ : 1 ≤ m ↦ (⊤ : ℕ∞) ^ m = ⊤) (pow_one ⊤) _ _ n_pos
  intro m _ h
  simp only [pow_add, h, pow_one]
  exact WithTop.top_mul_top
```

**Comment by pitmonticone** [comment]
Ok, I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

**Comment by sgouezel** [comment]
In fact, this should be generalized to `WithTop`. Could you take the proof from `ENNReal.top_pow` and move it to `Algebra.Order.Ring.WithTop` at Line 170 (and then restate `ENNReal.top_pow` and `ENat.top_pow` as specialized versions of the general statement?)

**Comment by pitmonticone** [comment]
I should have addressed all the review requests: 

- [x] Add generalised version of `top_pow` in `Algebra/Order/Ring/WithTop`
- [x] Add specialised version of `top_pow` in `Data/ENat/Basic.lean`
- [x] Add specialised version of `top_pow` in `Data/ENNReal/Operations.lean` 

Thank you @sgouezel for your review.

**Comment by sgouezel** [comment]
My comment was to copy the proof of `ENNReal.top_pow` because it is better than the proof of the current PR. The comment still stands :-)

File: Mathlib/Algebra/Order/Ring/WithTop.lean [hunk 2]
```diff
@@ -167,6 +167,14 @@ instance instMonoidWithZero : MonoidWithZero (WithTop α) where
 
 @[simp, norm_cast] lemma coe_pow (a : α) (n : ℕ) : (↑(a ^ n) : WithTop α) = a ^ n := rfl
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : WithTop α)^n = ⊤ := by
+  apply Nat.le_induction (P := fun m : ℕ ↦ fun _ : 1 ≤ m ↦
+    (⊤ : WithTop α) ^ m = ⊤) (pow_one ⊤) _ _ n_pos
+  intro m _ h
+  simp only [pow_add, h, pow_one]
+  exact top_mul_top
+#align withtop.top_pow WithTop.top_pow
```
Comment (sgouezel) [inline Mathlib/Algebra/Order/Ring/WithTop.lean:176]: no #align for new statements. It's only useful for statements that were already existing in Lean 3, for porting tools.

File: Mathlib/Data/ENat/Basic.lean [hunk 3]
```diff
@@ -89,6 +89,9 @@ theorem coe_sub (m n : ℕ) : ↑(m - n) = (m - n : ℕ∞) :=
 @[simp] theorem mul_top (hm : m ≠ 0) : m * ⊤ = ⊤ := WithTop.mul_top hm
 @[simp] theorem top_mul (hm : m ≠ 0) : ⊤ * m = ⊤ := WithTop.top_mul hm
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℕ∞)^n = ⊤ := WithTop.top_pow n_pos
+#align enat.top_pow ENat.top_pow
```
Comment (sgouezel) [inline Mathlib/Data/ENat/Basic.lean:93]: ditto

File: Mathlib/Algebra/Order/Ring/WithTop.lean [hunk 4]
```diff
@@ -167,6 +167,14 @@ instance instMonoidWithZero : MonoidWithZero (WithTop α) where
 
 @[simp, norm_cast] lemma coe_pow (a : α) (n : ℕ) : (↑(a ^ n) : WithTop α) = a ^ n := rfl
 
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : WithTop α)^n = ⊤ := by
```
Comment (sgouezel) [inline Mathlib/Algebra/Order/Ring/WithTop.lean:170]:
```suggestion
theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : WithTop α) ^ n = ⊤ := by
```
Spaces in power operation, here and elsewhere in the PR.

File: Mathlib/Data/ENNReal/Operations.lean [hunk 5]
```diff
@@ -219,10 +219,7 @@ theorem top_mul_top : ∞ * ∞ = ∞ := WithTop.top_mul_top
 #align ennreal.top_mul_top ENNReal.top_mul_top
 
 -- Porting note (#11215): TODO: assume `n ≠ 0` instead of `0 < n`
--- Porting note (#11215): TODO: generalize to `WithTop`
-theorem top_pow {n : ℕ} (h : 0 < n) : ∞ ^ n = ∞ :=
-  Nat.le_induction (pow_one _) (fun m _ hm => by rw [pow_succ, hm, top_mul_top]) _
-    (Nat.succ_le_of_lt h)
+theorem top_pow {n : ℕ} (n_pos : 0 < n) : (⊤ : ℝ≥0∞)^n = ⊤ := WithTop.top_pow n_pos
```
Comment (sgouezel) [inline Mathlib/Data/ENNReal/Operations.lean:222]: In ENNReal, the notation ∞ is preferred.

**Comment by pitmonticone** [comment]
Ok, I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

File: Mathlib/Data/ENat/Basic.lean (hunk 3 above)
Comment (sgouezel) [reply Mathlib/Data/ENat/Basic.lean:93]: The "no #align for new statements" also applies here. That's the meaning of "ditto".

Comment (pitmonticone) [reply Mathlib/Data/ENat/Basic.lean]: No, I did not... sorry. 

Done now.

**Comment by pitmonticone** [comment]
Now I should have addressed all the review requests. 

Thank you very much @sgouezel for your review!

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**